src/placementEngine.js

```diff
@@ -1,42 +1,135 @@
-  // A function that emits a `close` event on the bus.
-  const onClose = promo => bus.next({ type: 'close', promo })
+/**
+ * Increments the counter and sets the timestamp for the given entity.
+ *
+ * @private
+ */
+function incrementCounter (keyPath, ts) {
+  // Return the identity function (NOOP) if the last element in the key path
+  // is falsey.
+  if (!last(keyPath)) return id
+
+  return update(keyPath, state => {
+    const { count } = state || { count: 0 }
+    return {
+      count: count + 1,
+      timestamp: ts
+    }
+  })
+}
+
+/**
+ * Creates a new placement context that contains objects and functions to be
+ * made available to the constraint queries.
+ *
+ * @private
+ */
+function createContext (userAgent, user) {
+  const uaParser = new UAParser(userAgent)
+
+  return {
+    // Objects
+    browser: uaParser.getBrowser(),
+    device: uaParser.getDevice(),
+    os: uaParser.getOS(),
+    user,
+    window,
+
+    // Functions
+    age,
+    lower: toLower,
+    upper: toUpper
+  }
+}
+
+/**
+ * The placement engine is responsible for placing the given promos. When the
+ * engine is run, the placed promos are returned along with various callbacks
+ * for the calling application to call when events occur (i.e. a promo is
+ * clicked, etc).
+ *
+ * The placement engine returns a promise that resolves to an object containing
+ * the placed promos and the callback functions.
+ *
+ * @param {Storage} storage The storage object.
+ * @param {String} userAgent The user agent string.
+ * @param {Array} promos The list of the candidate promos.
+ * @returns {Promise} A promise.
+ */
+export default function placementEngine ({ promos, storage, userAgent }) {
+  const updateUser = (f, user) => set(storage, f(user))
+  let user = updateUser(incrementVisits, get(storage))
 
-  // Create the initial state object.
-  const initialState = { user }
+  const onClick = promo => {
+    const f = trackEngagement(promo, timestamp())
+    updateUser(f, user)
+  }
 
-  // The state signal emits the current placement engine state whenever an
-  // event is emitted on the bus.
-  const stateSignal = bus
-    // Emit an initial `visit` event on the bus.
-    .startWith({ type: 'visit' })
+  const onClose = promo => {
+    const f = blockPromo(promo, timestamp())
+    updateUser(f, user)
+  }
 
-    // Run the state machine function over the events emitted on the bus.
-    .stateMachine(stateMachine(promos, window), initialState)
+  const onView = promo => {
+    const f = trackImpression(promo, timestamp())
+    updateUser(f, user)
+  }
 
-    // Emit the placed promos and callback functions.
-    .map(({ promos }) => ({ promos, onClick, onClose }))
+  // Place the promos.
+  const context = createContext(userAgent, user)
+  const placedPromos = placePromos(context, promos)
 
-  return stateSignal
+  // Return a promise containing the placed promos, and the callback functions.
+  return Promise.resolve({ promos: placedPromos, onClick, onClose, onView })
```

Comment:
I suppose we might want to reject if some unrecoverable error happens, but this seems fine for now. It's at least a much nicer API with flexibility for the future.